File: rust-sdk/src/agent.rs
//! Agent trait definition.

use std::future::Future;

use tokio::sync::mpsc;

use crate::types::{AgentOutput, Inputs};

/// A sender for streaming log messages during agent execution.
#[derive(Clone)]
pub struct LogSender {
    tx: mpsc::Sender<String>,
}

impl LogSender {
    pub(crate) fn new(tx: mpsc::Sender<String>) -> Self {
        Self { tx }
    }

    /// Send a log message.
    pub async fn send(&self, msg: impl Into<String>) {
        let _ = self.tx.send(msg.into()).await;
    }

    /// Send a log message (non-async, best-effort).
    pub fn try_send(&self, msg: impl Into<String>) {
        let _ = self.tx.try_send(msg.into());
    }
}

/// Trait for implementing an Actflow agent.
///
/// Implement this trait to create your own agent that can be executed
/// by the Actflow workflow engine.
///
/// # Example
///
/// ```rust,ignore
/// use actflow_agent_sdk::{Agent, AgentOutput, Inputs, LogSender};
///
/// struct MyAgent;
///
/// impl Agent for MyAgent {
///     async fn run(
///         &self,
///         pid: String,
///         nid: String,
///         inputs: Inputs,
///         log: LogSender,
///     ) -> AgentOutput {
///         log.send("Processing...").await;
///         AgentOutput::success(serde_json::json!({"result": "done"}))
///     }
/// }
/// ```
pub trait Agent: Send + Sync + 'static {
    /// Execute the agent logic.
    ///
    /// # Arguments
    ///
    /// * `pid` - Process ID
    /// * `nid` - Node ID in the workflow
    /// * `inputs` - Input data for this agent
    /// * `log` - Log sender for streaming log messages
    ///
    /// # Returns
    ///
    /// Returns an `AgentOutput` containing the execution status and output data.
    fn run(
        &self,
        pid: String,
        nid: String,
        inputs: Inputs,
        log: LogSender,
    ) -> impl Future<Output = AgentOutput> + Send;

    /// Called when the agent should shut down.
    ///
    /// Override this method to perform cleanup operations.
    /// Default implementation does nothing.
    fn shutdown(&self) -> impl Future<Output = ()> + Send {
        async {}
    }
}

File: go-sdk/agent.go
// Package sdk provides a Go SDK for building Actflow agents.
package sdk

import "context"

// Agent is the interface that users must implement to create an agent.
type Agent interface {
	// Run executes the agent logic.
	// The log channel is used to send log messages during execution.
	// Close the log channel or return to complete execution.
	Run(ctx context.Context, pid, nid string, inputs map[string]any, log chan<- string) (*Output, error)

	// Shutdown is called when the agent should shut down.
	// Use this to perform cleanup operations.
	Shutdown(ctx context.Context) error
}

// ExecutionStatus represents the status of agent execution.
type ExecutionStatus int32

const (
	StatusPending   ExecutionStatus = 0
	StatusSucceeded ExecutionStatus = 1
	StatusFailed    ExecutionStatus = 2
	StatusException ExecutionStatus = 3
	StatusStopped   ExecutionStatus = 4
	StatusPaused    ExecutionStatus = 5
)

// String returns the string representation of the status.
func (s ExecutionStatus) String() string {
	switch s {
	case StatusPending:
		return "PENDING"
	case StatusSucceeded:
		return "SUCCEEDED"
	case StatusFailed:
		return "FAILED"
	case StatusException:
		return "EXCEPTION"
	case StatusStopped:
		return "STOPPED"
	case StatusPaused:
		return "PAUSED"
	default:
		return "UNKNOWN"
	}
}

// Output represents the result of agent execution.
type Output struct {
	// Status is the execution status.
	Status ExecutionStatus
	// Outputs contains the output data.
	Outputs any
	// Error message (when Status is Failed).
	Error string
	// Exception message (when Status is Exception).
	Exception string
}

// Success creates a successful output.
func Success(outputs any) *Output {
	return &Output{
		Status:  StatusSucceeded,
		Outputs: outputs,
	}
}

// Failed creates a failed output with an error message.
func Failed(err string) *Output {
	return &Output{
		Status: StatusFailed,
		Error:  err,
	}
}

// FailedErr creates a failed output from an error.
func FailedErr(err error) *Output {
	return &Output{
		Status: StatusFailed,
		Error:  err.Error(),
	}
}

// Exception creates an exception output.
func Exception(exception string) *Output {
	return &Output{
		Status:    StatusException,
		Exception: exception,
	}
}
